File: new_version_with_notes/practice/python/ops/bn.py
def BatchNorm(img, mean, var, gamma, bias):
    """
    对图像进行批量归一化处理。

    参数:
    img (numpy.ndarray): 输入图像。
    mean (numpy.ndarray): 归一化处理所用的均值。
    var (numpy.ndarray): 归一化处理所用的方差。
    gamma (numpy.ndarray): 缩放参数。
    bias (numpy.ndarray): 偏移参数。

    返回:
    numpy.ndarray: 批量归一化后的图像。
    """
    # 获取图像的高度、宽度和通道数
    h = img.shape[0]
    w = img.shape[1]
    c = img.shape[2]

    # 对每个通道进行批量归一化处理
    for c_ in range(c):
        # 提取单个通道的数据
        data = img[:, :, c_]
        # 执行批量归一化：(data - mean) / sqrt(var + epsilon)
        data_ = (data - mean[c_]) / (pow(var[c_] + 1e-5, 0.5))
        # 应用缩放和偏移
        data_ = data_ * gamma[c_]
        data_ = data_ + bias[c_]
        # 更新图像的该通道数据
        img[:, :, c_] = data_

    return img  # 返回批量归一化后的图像

File: practice/cpp/4th_no_malloc/ops/bn.cc
#include <immintrin.h>

#include <cmath>
#include <cstdint>
#include <iostream>
#include <opencv2/opencv.hpp>
#include <string>

// 定义预加载版本的批量归一化函数，此版本不执行任何操作
void MyBatchNormPreLoad(void* img_in,
                        void* img_out,
                        float* mean,
                        float* var,
                        float* gamma,
                        float* bias,
                        int h,
                        int w,
                        int c) {
  return;
}

// 批量归一化函数，执行实际的归一化操作
void MyBatchNorm(void* img_in,
                 void* img_out,
                 float* mean,
                 float* var,
                 float* gamma,
                 float* bias,
                 int h,
                 int w,
                 int c) {
  // 类型转换输入和输出为 float 指针
  float* img = (float*)img_in;
  float* out = (float*)img_out;

  // 遍历每个通道
  for (auto c_ = 0; c_ < c; c_++) {
    // 获取当前通道的均值、方差、比例因子和偏置
    auto m = mean[c_];
    auto v = var[c_];
    auto gm = gamma[c_];
    auto bi = bias[c_];

    // 遍历每个像素
    for (auto hw = 0; hw < h * w; hw++) {
      // 读取当前像素的值
      auto data = img[hw * c + c_];
      // 执行归一化操作：(data - mean) / sqrt(variance + epsilon)
      auto data_ = (data - m) / sqrt(v + 1e-5);
      // 应用比例因子和偏置
      data_ = data_ * gm + bi;
      // 将归一化结果写入输出数组
      out[hw * c + c_] = data_;
    }
  }
}